types/wlr_text_input.c

```diff
@@ -0,0 +1,182 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <stddef.h>
+
+#include <wlr/types/wlr_seat.h>
+#include <wlr/types/wlr_text_input.h>
+
+#include <wlr/util/log.h>
+
+#include "text-input-unstable-v3-protocol.h"
+#include "util/signal.h"
+
+
+static const struct zwp_text_input_v3_interface text_input_impl;
+
+static struct wlr_text_input *text_input_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_text_input_v3_interface,
+		&text_input_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+void wlr_text_input_send_enter(struct wlr_text_input *text_input,
+		struct wlr_surface *wlr_surface) {
+	zwp_text_input_v3_send_enter(text_input->resource,
+		text_input->next_serial++, wlr_surface->resource);
+}
+
+void wlr_text_input_send_leave(struct wlr_text_input *text_input,
+		struct wlr_surface *wlr_surface) {
+	zwp_text_input_v3_send_leave(text_input->resource,
+		text_input->next_serial++, wlr_surface->resource);
+}
+
+void wlr_text_input_send_preedit_string(struct wlr_text_input *text_input,
+		const char *text, uint32_t cursor) {
+	zwp_text_input_v3_send_preedit_string(text_input->resource, text, cursor);
+}
+
+void wlr_text_input_send_commit_string(struct wlr_text_input *text_input,
+		const char *text) {
+	zwp_text_input_v3_send_commit_string(text_input->resource, text);
+}
+
+static void text_input_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wlr_log(L_INFO, "STUB: text_input_destroy");
+}
+
+static void text_input_enable(struct wl_client *client,
+		struct wl_resource *resource, uint32_t serial,
+		uint32_t show_input_panel) {
+	struct wlr_text_input *text_input = text_input_from_resource(resource);
+	if (serial >= text_input->next_serial) {
+		wlr_log(L_INFO,
+			"text_input_enable called with a serial that wasn't sent yet: %d",
+			serial);
+	}
+	wlr_signal_emit_safe(&text_input->events.enable_notify,
+		(void*)&show_input_panel);
+}
+
+static void text_input_disable(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_text_input *text_input = text_input_from_resource(resource);
+	wlr_signal_emit_safe(&text_input->events.disable_notify, text_input);
+}
+
+static void text_input_set_surrounding_text(struct wl_client *client,
+		struct wl_resource *resource, const char *text, int32_t cursor,
+		int32_t anchor) {
+	wlr_log(L_INFO, "STUB: set_surrounding_text(%s, %d, %d)", text, cursor,
+		anchor);
+}
+
+static void text_input_set_content_type(struct wl_client *client,
+		struct wl_resource *resource, uint32_t hint, uint32_t purpose) {
+	wlr_log(L_INFO, "STUB: set_content_type(%d, %d)", hint, purpose);
+}
+
+static void text_input_set_cursor_rectangle(struct wl_client *client,
+		struct wl_resource *resource, int32_t x, int32_t y, int32_t width,
+		int32_t height) {
+	wlr_log(L_INFO, "STUB: set_cursor_rectangle(%d, %d, %d, %d)", x, y, width,
+		height);
```

Comment:
This stuff too, you can just stash it and not use it on rootston but definitely implement the state transitions